### PR TITLE
Gemma 4 E2B: GGUF INT4 direct conversion + QNN HTP lowering

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

Enables converting the **unsloth/gemma-4-E2B-it GGUF (Q4_K_M)** directly into a working INT4 ONNX model that targets the **QNN EP / Hexagon HTP** — reusing the GGUF's existing INT4 weights instead of running expensive Olive quantization. Pairs with olive-recipes#432 (adds a GGUF-direct `config_gguf.json`).

## What's here

**1. Gemma 4 GGUF INT4 conversion fixes** (make `mobius build-gguf --keep-quantized` work for E2B):
- Collapse per-layer `feed_forward_length` array → scalar `intermediate_size` + `use_double_wide_mlp`.
- Map the top-level per-layer-input tensors + strip the `layer_scalar.weight` artefact.
- Thread the INT4 `MatMulNBits` linear class through Gemma4 (`--keep-quantized` was silently producing a **float** model) + tie the float LM head.
- Fix KV-shared standard-Attention shape inference (opset-24 `Attention` present outputs are un-inferred → `o_proj` collapsed → model unloadable).
- Fix the Gemma GGUF activation default (`silu` → `gelu_pytorch_tanh`; the wrong default produced garbage output).

**2. QNN HTP lowering** (two EP-capability-gated transforms; both default-on, off for `qnn`):
- `supports_attention=False` → `DecomposeAttentionPass` rewrites the fused opset-24 `Attention` op into SDPA primitives (Reshape/Transpose/MatMul/Softmax/Add, Tile for GQA). An `onnx_ir` InPlacePass so it can rewire the 3 outputs (present-KV are graph outputs on some layers, dead on shared-KV layers — mixed arity the pattern rewriter can't express).
- `supports_matmul_nbits=False` → `com.microsoft::MatMulNBits` lowered to blocked-uint4 `DequantizeLinear` + `MatMul` (QDQ), implemented as a **model-local ir.Function** expanded by the existing InlinePass mechanism (like SkipLayerNormalization / PackedMultiHeadAttention). Supported EPs keep the native contrib op + kernel.

## Verification

- Dequantized GGUF weights match the gated `google/gemma-4-E2B-it` checkpoint (corr ≥ 0.98).
- Both lowerings are numerically identical to the originals via ORT (maxdiff 0.0 for prefill/decode/GQA/mask; ~5e-7 with softcap due to fp rounding).
- The INT4 model generates correct text on ORT ("capital of France → Paris", "capital of Japan → Tokyo", valid haiku).
- A full `qnn` build lowers to **0 Attention / 35 Softmax** and **0 MatMulNBits / 205 DequantizeLinear**, all 30 present-KV graph outputs preserved; CPU/CUDA builds unchanged (still emit GQA + MatMulNBits).
- Empirically confirmed on a Snapdragon X box (ORT 1.27, QNN SDK 2.48.40): HTP runs the primitive ops (MatMul/Softmax/RMSNormalization/Gelu) but **not** fused Attention/MatMulNBits — exactly what this lowering removes. (Plugin QNN EP is engaged via `add_provider_for_devices`, and HTP requires fully static shapes.)

## Remaining / follow-up

- Unit tests for `DecomposeAttentionPass` and the `MatMulNBits` function body (in progress).
- End-to-end HTP EPContext compile validation (olive-recipes `config_gguf.json` path) not yet hardware-validated.

🤖 Generated with GitHub Copilot CLI

### CODE REVIEW

**Review by rui-ren — APPROVED** [review]

**Comment by justinchuby** [comment]
### Update: unit tests added + HTP execution validated on-device

**Tests** (commit `f74b017`):
- `_decompose_attention_test.py` — numerical parity vs the fused opset-24 Attention op across prefill / decode-with-past / MHA / GQA / softcap / mask-only, KV-cache graph-output rewiring, rank-4 left-untouched, and EP gating.
- `matmul_nbits_test.py` — the inlined QDQ function body matches the native `MatMulNBits` op bit-for-bit (incl. odd n_blocks), plus EP gating.
- 303 tests pass across functions / rewrite_rules / gguf suites; ruff clean.

**On-device HTP validation** (Snapdragon X, ORT 1.27, onnxruntime-qnn 2.4.0 / QNN SDK 2.48.40):
- Built a tiny `gemma4_text` for `qnn` (f16), pinned static input shapes, engaged the plugin QNN EP via `register_execution_provider_library` + `SessionOptions.add_provider_for_devices(npu)`.
- The **lowered model compiles onto the Hexagon HTP** (QNN graph-preparation stages run) and produces finite logits — confirming the decomposed attention + QDQ actually offload to the NPU. Standalone probes confirmed the fused `Attention` op and `MatMulNBits` do **not** run on HTP (they're what this lowering removes); `MatMul` / `Softmax` / `RMSNormalization` / `Gelu` / `DequantizeLinear` all do.
- Filed the upstream gap as onnxruntime/onnxruntime-qnn#646 (QNN EP lacks a standard ONNX `Attention` kernel).

A few index/control ops (`CumSum` / `Range` / int64 gather) still fall to CPU under fallback — a follow-up, not a blocker for the attention/matmul compute offload.

**Comment by justinchuby** [comment]
### Session update: build-gguf `--static-cache` + QNN HTP findings

**New this branch (commit `84fcf06`):** `build-gguf` now accepts `--static-cache` / `--max-seq-len`, threading a static KV cache (pre-allocated fixed-width buffers written in place via `TensorScatter`) through `build_from_gguf` → `CausalLMTask`. This produces a fully static-shaped graph as required by fixed-shape runtimes such as the QNN HTP backend. Added `TestBuildGgufStaticCache` (fixed-width `key_cache`/`value_cache` I/O, no dynamic `past_key_values.*`; explicit-task rejection). All gguf-builder + CLI tests pass; ruff clean; no new mypy errors.

**QNN HTP experiment findings (Snapdragon X, ORT 1.27, onnxruntime-qnn 2.4.0):**
- `--ep qnn` lowering verified: the GGUF INT4 E2B build decomposes the fused opset-24 `Attention` and inlines `MatMulNBits` → QDQ (0 `Attention`/`MatMulNBits` nodes remain; 205 `DequantizeLinear` + nibble-unpack).
- **Full 35-layer / 4.65B E2B fails single-graph HTP compose on size** (`EP_FAIL: Failed to compose Qnn graph`). Smaller subsets compose fine → needs model splitting + weight-shared EPContext binaries.
- Olive `StaticLLM` QNN path assumes the ORT-GenAI static-KV contract (`past_seq_len`/`total_seq_len` + sliding-window KV); mobius's decomposed standard-attention output doesn't match.

**Top remaining prerequisite:** gemma4 does **not** implement the `StaticCacheState` dispatch (KV-shared layers + sliding/full alternation + dual local/global RoPE), so `--static-cache` currently raises `TypeError` for gemma4. This is correctness-critical surgery on the most complex model in mobius and needs HF numerical-parity validation — intentionally not rushed unsupervised.

CPU tok/s baselines to beat: llama.cpp 43.7, mobius ONNX (ORT CPU) 18.2, HF fp32 0.04.

Recipe docs (olive-recipes PR #432) updated to match.

**Comment by justinchuby** [comment]
### gemma4 static KV cache (commit `5d804b7`) + HTP findings

**Implemented & verified** — `Gemma4TextAttention`/`Gemma4DecoderLayer` now support `CausalLMTask(static_cache=True)` (previously rejected). Handles gemma4's specifics: KV-shared layers borrow the source layer's static buffer (no own cache), dual head_dim (sliding 256 / full 512) via `Gemma4TextModel.static_kv_cache_specs()`, and a static-cache attention bias for both layer types.

**Correctness:** static-cache decode logits match the dynamic reference at **every position (~1e-6 fp32)** across sliding-only, dual-head_dim, and KV-shared configs; and the **full E2B static model generates correct text on ORT CPU**: `build-gguf "...gemma-4-E2B-it-Q4_K_M.gguf" --keep-quantized --ep qnn --dtype f16 --static-cache --max-seq-len 512` → *"The capital of France is **Paris**."* The graph is fully static + decomposed + QDQ (0 fused Attention, 0 MatMulNBits, 30 TensorScatter, 15 cache layers with dual head_dim). Added 3 build-graph tests; all gemma4/task/static tests pass; ruff clean; no new mypy errors.

> ⚠️ Note (debugging gotcha, now in repo memory): mobius emits the attention mask value `dtype.min` (−3.4e38) as a **graph FLOAT initializer**. Test harnesses that randomly fill *all* float initializers corrupt the mask and report spurious per-position mismatches — skip `|value|>1e30` initializers.

**HTP status (Snapdragon X, onnxruntime-qnn 2.4.0):** the E2B still needs splitting for HTP:
1. the 4.7 GB per-layer embedding overflows QNN's 32-bit static-tensor size → keep embeddings on CPU (splitting after the embedding Gathers gives a 2.15 GB transformer that *composes* on HTP);
2. the 35-layer transformer fails single-graph HTP *finalization* (~20 min) — a 4-layer chunk composes+runs on HTP in ~21 s → needs weight-shared EPContext chunking;
3. perf: the decompose+inline-QDQ form runs slowly on HTP (~0.2 tok/s, ops fall to CPU) — a reasonable NPU number needs Olive `MatMulNBitsToQDQ` (clean int4, no runtime nibble-unpack) + `OnnxStaticQuantization` (int activations) + chunked EPContext, now unblocked by this static-cache support.

Recipe docs updated in olive-recipes PR #432.

**Comment by justinchuby** [comment]
### Full HTP op offload via mobius qnn lowering (commit `be98116`)

Changed the mobius QNN lowering so a static-cache decoder runs **fully on the Hexagon HTP** (no CPU fallback). Empirically probed (Snapdragon X, onnxruntime-qnn 2.4.0, `session.disable_cpu_ep_fallback=1`) which ops the HTP has no kernel for — each forced its node to CPU — and lowered them all:

| CPU-forced op | Lowered to |
|---|---|
| fused `Attention` | SDPA primitives (already) |
| `RotaryEmbedding` (opset-24) | rotate-half (Reshape/Slice/Mul/Sub/Add/Concat) — **new** `DecomposeRotaryEmbedding` |
| `TensorScatter` (static cache) | `ScatterND` (batch=1) — **new** `TensorScatterToScatterND` |
| `Tile` (GQA head-repeat) | `Expand` |
| `Range` (bias) | `Constant(arange)` + `Slice` |
| `MatMulNBits` | QDQ (already); nibble-unpack `BitwiseAnd`/`BitShift` constant-folds away |

Two new EP capability flags (`supports_rotary_embedding`, `supports_tensor_scatter`, both `False` for qnn) gate the new rules. All replacements are **numerically identical** (qnn-lowered gemma4 static model matches the fused reference at **maxdiff 0.0**). Added parity + op-replacement unit tests for both rules and a gemma4 build test asserting the qnn graph has no `RotaryEmbedding`/`TensorScatter`/`Tile`/`Range`.

**Result:** a non-quantized (fp16) gemma4 static model loads on QNN HTP with `disable_cpu_ep_fallback=1` and **composes + runs entirely on the NPU**.

**Remaining INT4 constraint (not a mobius rule):** weight-only `DequantizeLinear`→`MatMul` (float activations) runs on HTP only for *per-tensor* weights; *per-channel/blocked* int8/int4 weight DQ is CPU-forced unless the **activations are also quantized** (full int QDQ: DQ(act)+DQ(weight per-channel)→MatMul→Q, verified HTP-OK). So GGUF INT4 weights need Olive `OnnxStaticQuantization` (activation calibration) to keep their matmuls on HTP. All checks pass; ruff clean; no new mypy errors. Recipe docs updated (olive-recipes PR #432).

File: src/mobius/components/_common.py [hunk 1]
```diff
@@ -335,23 +335,29 @@ def create_static_cache_attention_bias(
...
-    zero_scalar = op.Constant(value_int=0)
-    one_scalar = op.Constant(value_int=1)
+    # KV positions: dense cache slot ids 0 .. max_seq_len - 1 -> (max_seq_len,).
+    # Emitted as a constant (not Range) because max_seq_len is a concrete int
+    # and the QNN HTP backend has no Range kernel (Range forces the node onto
+    # CPU). q_offsets is then a Slice of this constant to length S_q — both
+    # Constant and Slice run on HTP.
+    kv_slots = op.Constant(
+        value=ir.tensor(np.arange(max_seq_len, dtype=np.int64))
+    )  # (max_seq_len,)
```
Comment (justinchuby) [inline src/mobius/components/_common.py]: Should this be enabled only when static cache is used? @copilot please investigate

**Comment by justinchuby** [comment]
Addressed the Copilot review in c5a1db2:

- Build GQA `Expand` targets from runtime B/Nkv/S/H dimensions.
- Preserve `nonpad_kv_seqlen` via valid-prefix score masking.
- Use constant aranges sliced to runtime lengths for statically bounded KV caches, so QNN static-cache lowering emits no `Range`.
- Reject TensorScatter lowering for statically known batch > 1.
- Require rank-3 cos/sin before RotaryEmbedding decomposition.
- Corrected `Tile` references to `Expand`.
- The earlier EP-gated static-cache bias comment was already resolved in 7c49b8b.

Validation: lint clean; targeted lowering tests 22 passed; full suite 4718 passed, 334 skipped, 62 xfailed, with only the 11 known NeMo failures (#428). The same suite with NeMo ignored passes completely.
